### PR TITLE
const support for Nullable<>

### PR DESCRIPTION
**Pull request by krlmlr**
cargo-culted from many other implementations of `operator SEXP() const`

This came up when trying to use `const Nullable<...>&` in https://github.com/rstats-db/RMySQL/pull/90/files#r49576837

Seemed too minor to open an issue first. What would a test for this look like?

CC @peternowee

### CODE REVIEW

**Comment by eddelbuettel** [comment]
@dcdillon Care to look at this (and tests) ?

**Comment by dcdillon** [comment]
I will take a peek soon.  May not have time until the weekend (depends on the wife).

File: inst/unitTests/cpp/misc.cpp [hunk 1]
```diff
@@ -179,22 +179,22 @@ void test_stop_variadic() {
 }
 
 // [[Rcpp::export]]
-bool testNullableForNull(Nullable<NumericMatrix> M) {
+bool testNullableForNull(const Nullable<NumericMatrix>& M) {
```
Comment (eddelbuettel) [inline inst/unitTests/cpp/misc.cpp]: Methinks we may want these as `const` variants (post PR) as well as non-`const` variants just to be sure.

File: inst/include/Rcpp/Nullable.h [hunk 2]
```diff
@@ -76,9 +76,8 @@ namespace Rcpp {
          *
          * @throw 'not initialized' if object has not been set
          */
-        inline operator SEXP() {
-            checkIfSet();
```
Comment (eddelbuettel) [inline inst/include/Rcpp/Nullable.h]: Why did you remove `checkIfSet()` ?

Comment (krlmlr) [inline inst/include/Rcpp/Nullable.h]: Now forwarded to get().

Comment (eddelbuettel) [inline inst/include/Rcpp/Nullable.h]: I can see that.  Let me rephrase:  Why did you make that change? What is the reasoning?

I see no relation in that and the `const` change.  So why mess with it?

File: inst/unitTests/cpp/misc.cpp (hunk 1 above)
Comment (krlmlr) [inline inst/unitTests/cpp/misc.cpp]: It's no big deal to re-add the pre-PR versions, but I wonder under what circumstances code can succeed for `const&` but fail for a non-const object?

**Comment by dcdillon** [comment]
The question I have is with regards to the return type of SEXP of the `get()` function.  It doesn't provide any safety...so what is the use case of a `const Nullable<T>` object?

File: inst/include/Rcpp/Nullable.h (hunk 2 above)
Comment (krlmlr) [inline inst/include/Rcpp/Nullable.h]: You're right, it's an unrelated change. I included it, because it affects code in close proximity to the code that I was touching anyway. The reason for that change: in the unlikely event that the implementation of get() changes, you will also want to change the implementation of operator SEXP(), because they seem synonymous.

That said, I'm totally fine with excluding that particular change from this PR, if you prefer that.

**Comment by eddelbuettel** [comment]
As for your motivation in https://github.com/rstats-db/RMySQL/pull/90/files#r49576837, why not just initialize with empty strings a la `std::string foo = ""`.   Why insinst on `Nullable` and require this change?

**Comment by dcdillon** [comment]
What I would propose is adding an `T as()` function and a `T as() const` function to `Nullable` where the first one simply returns a `T(m_sexp)` and the second one returns a "make deep copy of SEXP and then create a `T` out of it" and return it.  This way we get actual const safety out of a `const Nullable<T>`.  It means that the efficiency of `T as() const` function would be really crappy, but I see no other way to make `const Nullable<T>` provide any safety at all.

**Comment by krlmlr** [comment]
The strings are passed to a C API which makes a difference between `NULL` and `""`. `Rcpp::Nullable` seemed handy.

**Comment by eddelbuettel** [comment]
@dcdillon I like that.  May look into what `Rcpp::clone()` does.

@krlmlr The idiom `std::string foo = "hello, world"; const char bar = foo.c_str();` is about as old `std::string` and gets you a C API string.  Again, no motivation for `Nullable` yet.

**Comment by krlmlr** [comment]
In that C API, passing `NULL` means "get your data elsewhere", while passing `""` means "this is my data, it is an empty string". How do I encapsulate that difference in a `std::string`?

**Comment by eddelbuettel** [comment]
Right. `std::string` alone can't do that as it can't hold a `NULL` (pointer).

**Comment by eddelbuettel** [comment]
Going to merge this as it should be of limited impact in its current form.

**Comment by kevinushey** [comment]
Also agree that the conversion operator should be `const`. :+1:

**Comment by dcdillon** [comment]
@kevinushey why should something that returns a value that allows you to modify the contents be `const`?  I know it's done as a convenience across Rcpp, but it's bastardizing `const` pretty horribly.  That is to say, if you can get a `SEXP` from a `const` function, the world is your oyster.  This would seem to go against any "traditional" meaning of `const`.

**Comment by eddelbuettel** [comment]
@dcdillon "This is R. Everything is bastardized."

We are also cheating somewhat with function interfaces like `double foo(const & NumericVector x)` because we cannot really undo the P in the `SEXP`.  It is what it is, but we all feel it is better to adhere to, and converge to, standard idioms.  As such, no harm in `const` qualifiers.

**Comment by dcdillon** [comment]
Yeah...i was more just wondering the logic, not arguing with it.
